torchtitan/components/tokenizer.py

```diff
@@ -25,3 +31,395 @@ def decode(self, *args, **kwargs) -> str:
+
+        add_bos = kwargs.get("add_bos", self.default_add_bos)
+        add_eos = kwargs.get("add_eos", self.default_add_eos)
+
+        # Get base token IDs from the underlying tokenizer
+        token_ids = self.tokenizer.encode(text).ids
+
+        # Add BOS token if requested and not already added by tokenizer
+        if not self.hf_adds_bos and add_bos:
+            if self.bos_id is not None:
+                token_ids.insert(0, self.bos_id)
+
+        # Add EOS token if requested and not already added by tokenizer
+        if not self.hf_adds_eos and add_eos:
+            if self.eos_id is not None:
+                token_ids.append(self.eos_id)
+
+        return token_ids
+
+    @override
+    def decode(self, *args, **kwargs) -> str:
+        """
+        Decode token IDs back to text.
+
+        Args:
+            token_ids (list[int]): List of token IDs to decode
+            **kwargs: Additional arguments passed to the underlying tokenizer's decode method
+                     (e.g., skip_special_tokens)
+
+        Returns:
+            str: Decoded text
+        """
+        # Extract token_ids from arguments
+        if len(args) >= 1:
+            token_ids = args[0]
+            # Pass through remaining kwargs
+            return self.tokenizer.decode(token_ids, **kwargs)
+        else:
+            token_ids = kwargs.pop("token_ids", [])
+            # Pass through remaining kwargs after removing token_ids
+            return self.tokenizer.decode(token_ids, **kwargs)
+
+    @property
+    def vocab_size(self) -> int:
+        """Get the vocabulary size."""
+        return len(self.tokenizer.get_vocab())
+
+    def get_vocab_size(self) -> int:
+        """Get the vocabulary size."""
+        return len(self.tokenizer.get_vocab())
+
+    def get_vocab(self) -> dict[str, int]:
+        """Get the vocabulary as a dictionary."""
+        return self.tokenizer.get_vocab()
+
+    def token_to_id(self, token: str) -> Optional[int]:
+        """Convert token to ID."""
+        return self.tokenizer.token_to_id(token)
+
+    def id_to_token(self, token_id: int) -> Optional[str]:
+        """Convert ID to token."""
+        return self.tokenizer.id_to_token(token_id)
+
+
+def build_hf_tokenizer(tokenizer_path: str) -> HuggingFaceTokenizer:
+    """
+    Builds a HuggingFaceTokenizer from the specified path.
+
+    This function creates a HuggingFaceTokenizer instance that handles BOS/EOS token
+    inference and intelligent encoding. The tokenizer automatically detects and loads
+    from various file formats and infers special token behavior.
+
+    Args:
+        tokenizer_path (str): Path to the directory containing tokenizer files.
+                             Should contain one or more of the supported file types.
+
+    Returns:
+        tokenizer (HuggingFaceTokenizer): Loaded tokenizer instance with intelligent BOS/EOS handling
+    """
+    tokenizer = HuggingFaceTokenizer(tokenizer_path)
```

Comment:
Great work! I would change FLUX tokenizer to our tokenizer after this diff landed